### PR TITLE
Add `--diagnostics` flag to write process context data to a file

### PR DESCRIPTION
**Pull request by yannbf**
This PR introduces a new flag to the CLI called `--report` which will write a log file called `chromatic-report.json` containing pretty useful information:

image

Will be useful in case we want to implement a fluid workflow with the storybook test-runner in CI.

Thanks @ghengeveld for all the help! <3

### CODE REVIEW

File: bin-src/lib/writeChromaticReport.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+// Extract important information from ctx, sort it and output into a json file
+export async function writeChromaticReport(ctx) {
+  const chromaticReportPath = 'chromatic-report.json';
+  const { argv, client, env, help, http, log, pkg, task, title, ...restCtx } = ctx;
+  const reportContext = Object.keys(restCtx)
+    .sort((a, b) => a.localeCompare(b))
+    .reduce((acc, key) => ({ ...acc, [key]: restCtx[key] }), {});
+  return writeFile(chromaticReportPath, reportContext, { spaces: 2 })
+    .then(() => ctx.log.info(wroteReport(chromaticReportPath, 'Chromatic')))
```
Comment (shilman) [inline bin-src/lib/writeChromaticReport.js:14]: why use chain-style promises if it's in an `async` function?

Comment (yannbf) [reply bin-src/lib/writeChromaticReport.js:14]: To catch errors without using try catch directly

**Comment by tmeasday** [comment]
Should we use a more useful name than `isolatorUrl`? This name is sort of a historical oddity :)

File: bin-src/lib/writeChromaticReport.js (hunk 1 above)
Comment (tmeasday) [reply bin-src/lib/writeChromaticReport.js:14]: What does everyone hate `try / catch` ? ;)

Comment (yannbf) [reply bin-src/lib/writeChromaticReport.js:14]: I don't hate it, just didn't use there. I can change if you'd like!

**Comment by yannbf** [comment]
> Should we use a more useful name than `isolatorUrl`? This name is sort of a historical oddity :)

I'm just using the raw format of the context. I spoke to @ghengeveld and the idea was to already introduce the flag so I could continue on the test-runner work and then make another PR to do some transformations to the object. I don't object doing that in this PR as well. Whatever seems best!

**Comment by ghengeveld** [comment]
> Should we use a more useful name than `isolatorUrl`? This name is sort of a historical oddity :)

If we go down that road, we should really only be including a fixed set of fields rather than the entire `ctx` object with some exceptions. If this is going to be an API of sorts, then I suppose that is a good idea. Perhaps we should even have a JSON schema for it.

**Comment by tmeasday** [comment]
I'm OK with just a data dump until we've had a chance to consider a proper API. But we should probably figure it out before we release this for real.

File: bin-src/lib/writeChromaticReport.js (hunk 1 above)
Comment (tmeasday) [reply bin-src/lib/writeChromaticReport.js:14]: Personally, I prefer to not mix async + promise code as I find it confusing. I'm not sure we have a strict rule about it though (perhaps we should? IDK).

Comment (yannbf) [reply bin-src/lib/writeChromaticReport.js]: Alright changed! Let me know if there is anything else!

**Review by shilman — APPROVED** [review]
LGTM! Outside the scope of this PR probably, but shouldn't `wroteReport` be a named export?

**Comment by yannbf** [comment]
> LGTM! Outside the scope of this PR probably, but shouldn't `wroteReport` be a named export?

Thanks! I just followed the rest of the files for consistency. All under `ui/messages` are default exported

**Comment by ghengeveld** [comment]
I'm still not sold on the name of this flag. Perhaps `--diagnostics` would be better. To me, `report` sounds like a structured thing, while `diagnostics` has less guarantees attached.

**Comment by yannbf** [comment]
> I'm still not sold on the name of this flag. Perhaps `--diagnostics` would be better. To me, `report` sounds like a structured thing, while `diagnostics` has less guarantees attached.

sounds good to me

**Comment by yannbf** [comment]
> 

Changes are done! Can you check them?

File: action-src/main.ts [hunk 2]
```diff
@@ -130,6 +130,7 @@ async function run() {
     const onlyChanged = getInput('onlyChanged');
     const externals = getInput('externals');
     const doNotStart = getInput('doNotStart');
+    const report = getInput('report');
```
Comment (ghengeveld) [inline action-src/main.ts:133]:
```suggestion
    const diagnostics = getInput('diagnostics');
```

File: action-src/main.ts [hunk 3]
```diff
@@ -170,6 +171,7 @@ async function run() {
       storybookCa: maybe(storybookCa),
       fromCI: true,
       interactive: false,
+      report: maybe(report),
```
Comment (ghengeveld) [inline action-src/main.ts:174]:
```suggestion
      diagnostics: maybe(diagnostics),
```

File: action.yml [hunk 4]
```diff
@@ -42,6 +42,9 @@ inputs:
   doNotStart:
     description: 'Do not attempt to start or build; use if your Storybook is already running'
     required: false
+  report:
```
Comment (ghengeveld) [inline action.yml:45]:
```suggestion
  diagnostics:
```

File: bin-src/lib/parseArgs.js [hunk 5]
```diff
@@ -76,6 +77,7 @@ export default function parseArgs(argv) {
         junitReport: { type: 'string' },
         list: { type: 'boolean' },
         interactive: { type: 'boolean', default: true },
+        report: { type: 'boolean' },
```
Comment (ghengeveld) [inline bin-src/lib/parseArgs.js:80]:
```suggestion
        diagnostics: { type: 'boolean' },
```

File: bin-src/lib/writeChromaticDiagnostics.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+export async function writeChromaticDiagnostics(ctx) {
+  const chromaticDiagnosticsPath = 'chromatic-diagnostics.json';
+  const { argv, client, env, help, http, log, pkg, task, title, ...restCtx } = ctx;
+  const diagnosticsContext = Object.keys(restCtx)
+    .sort((a, b) => a.localeCompare(b))
+    .reduce((acc, key) => ({ ...acc, [key]: restCtx[key] }), {});
+  try {
+    await writeFile(chromaticDiagnosticsPath, diagnosticsContext, { spaces: 2 });
+
+    ctx.log.info(wroteReport(chromaticDiagnosticsPath, 'Chromatic'));
```
Comment (ghengeveld) [inline bin-src/lib/writeChromaticDiagnostics.js:16]:
```suggestion
    ctx.log.info(wroteReport(chromaticDiagnosticsPath, 'Chromatic diagnostics'));
```

File: bin-src/ui/messages/info/wroteReport.stories.js [hunk 7]
```diff
@@ -4,4 +4,5 @@ export default {
   title: 'CLI/Messages/Info',
 };
 
-export const WroteReport = () => wroteReport('./chromatic-test-report.xml');
+export const WroteDiagnostics = () => wroteReport('./chromatic-report.xml', 'Chromatic');
```
Comment (ghengeveld) [inline bin-src/ui/messages/info/wroteReport.stories.js:7]:
```suggestion
export const WroteDiagnosticsReport = () => wroteReport('./chromatic-diagnostics.xml', 'Chromatic diagnostics');
```

**Review by ghengeveld — APPROVED** [review]
Nice work. You missed a couple places though.

**Comment by yannbf** [comment]
closing in favor of #479

**Pull request closed without merging**